diff -up ./mozilla/security/nss/cmd/shlibsign/shlibsign.c.587634 ./mozilla/security/nss/cmd/shlibsign/shlibsign.c
--- ./mozilla/security/nss/cmd/shlibsign/shlibsign.c.587634	2014-08-07 08:06:22.753078548 -0700
+++ ./mozilla/security/nss/cmd/shlibsign/shlibsign.c	2014-08-07 08:08:18.013079755 -0700
@@ -703,6 +703,7 @@ int main(int argc, char **argv)
     int i;
     PRBool verify = PR_FALSE;
     static PRBool FIPSMODE = PR_FALSE;
+    PRBool successful = PR_FALSE;
 
 #ifdef USES_LINKS
     int ret;
@@ -1247,6 +1248,8 @@ int main(int argc, char **argv)
     }
 #endif
 
+    successful = PR_TRUE;
+
 cleanup:
     if (pFunctionList) {
         /* C_Finalize will automatically logout, close session, */
@@ -1283,5 +1286,8 @@ cleanup:
     }
     PR_Cleanup();
 
-    return crv;
+    if (crv != CKR_OK)
+       return crv;
+    
+    return (successful) ? 0 : 1;
 }
