diff -up ./mozilla/security/nss/cmd/lib/basicutil.c.conflicts ./mozilla/security/nss/cmd/lib/basicutil.c
--- ./mozilla/security/nss/cmd/lib/basicutil.c.conflicts	2013-09-12 17:15:16.988951612 -0700
+++ ./mozilla/security/nss/cmd/lib/basicutil.c	2013-09-12 17:16:29.709951489 -0700
@@ -26,7 +26,7 @@
 
 #include "secoid.h"
 
-extern long DER_GetInteger(SECItem *src);
+extern long DER_GetInteger(const SECItem *src);
 
 static PRBool wrapEnabled = PR_TRUE;
 
