diff --git a/cmd/bltest/tests/aes_gcm/aad0 b/cmd/bltest/tests/aes_gcm/aad1
rename from cmd/bltest/tests/aes_gcm/aad0
rename to cmd/bltest/tests/aes_gcm/aad1
diff --git a/cmd/bltest/tests/aes_gcm/aad8 b/cmd/bltest/tests/aes_gcm/aad10
copy from cmd/bltest/tests/aes_gcm/aad8
copy to cmd/bltest/tests/aes_gcm/aad10
diff --git a/cmd/bltest/tests/aes_gcm/aad9 b/cmd/bltest/tests/aes_gcm/aad11
copy from cmd/bltest/tests/aes_gcm/aad9
copy to cmd/bltest/tests/aes_gcm/aad11
diff --git a/cmd/bltest/tests/aes_gcm/aad10 b/cmd/bltest/tests/aes_gcm/aad13
copy from cmd/bltest/tests/aes_gcm/aad10
copy to cmd/bltest/tests/aes_gcm/aad13
diff --git a/cmd/bltest/tests/aes_gcm/aad11 b/cmd/bltest/tests/aes_gcm/aad14
copy from cmd/bltest/tests/aes_gcm/aad11
copy to cmd/bltest/tests/aes_gcm/aad14
diff --git a/cmd/bltest/tests/aes_gcm/aad12 b/cmd/bltest/tests/aes_gcm/aad15
rename from cmd/bltest/tests/aes_gcm/aad12
rename to cmd/bltest/tests/aes_gcm/aad15
diff --git a/cmd/bltest/tests/aes_gcm/aad13 b/cmd/bltest/tests/aes_gcm/aad16
copy from cmd/bltest/tests/aes_gcm/aad13
copy to cmd/bltest/tests/aes_gcm/aad16
diff --git a/cmd/bltest/tests/aes_gcm/aad14 b/cmd/bltest/tests/aes_gcm/aad17
copy from cmd/bltest/tests/aes_gcm/aad14
copy to cmd/bltest/tests/aes_gcm/aad17
diff --git a/cmd/bltest/tests/aes_gcm/aad1 b/cmd/bltest/tests/aes_gcm/aad2
copy from cmd/bltest/tests/aes_gcm/aad1
copy to cmd/bltest/tests/aes_gcm/aad2
diff --git a/cmd/bltest/tests/aes_gcm/aad2 b/cmd/bltest/tests/aes_gcm/aad3
copy from cmd/bltest/tests/aes_gcm/aad2
copy to cmd/bltest/tests/aes_gcm/aad3
diff --git a/cmd/bltest/tests/aes_gcm/aad3 b/cmd/bltest/tests/aes_gcm/aad4
copy from cmd/bltest/tests/aes_gcm/aad3
copy to cmd/bltest/tests/aes_gcm/aad4
diff --git a/cmd/bltest/tests/aes_gcm/aad4 b/cmd/bltest/tests/aes_gcm/aad5
copy from cmd/bltest/tests/aes_gcm/aad4
copy to cmd/bltest/tests/aes_gcm/aad5
diff --git a/cmd/bltest/tests/aes_gcm/aad5 b/cmd/bltest/tests/aes_gcm/aad7
copy from cmd/bltest/tests/aes_gcm/aad5
copy to cmd/bltest/tests/aes_gcm/aad7
diff --git a/cmd/bltest/tests/aes_gcm/aad6 b/cmd/bltest/tests/aes_gcm/aad8
rename from cmd/bltest/tests/aes_gcm/aad6
rename to cmd/bltest/tests/aes_gcm/aad8
diff --git a/cmd/bltest/tests/aes_gcm/aad7 b/cmd/bltest/tests/aes_gcm/aad9
copy from cmd/bltest/tests/aes_gcm/aad7
copy to cmd/bltest/tests/aes_gcm/aad9
diff --git a/cmd/bltest/tests/aes_gcm/ciphertext0 b/cmd/bltest/tests/aes_gcm/ciphertext1
rename from cmd/bltest/tests/aes_gcm/ciphertext0
rename to cmd/bltest/tests/aes_gcm/ciphertext1
diff --git a/cmd/bltest/tests/aes_gcm/ciphertext8 b/cmd/bltest/tests/aes_gcm/ciphertext10
copy from cmd/bltest/tests/aes_gcm/ciphertext8
copy to cmd/bltest/tests/aes_gcm/ciphertext10
diff --git a/cmd/bltest/tests/aes_gcm/ciphertext9 b/cmd/bltest/tests/aes_gcm/ciphertext11
copy from cmd/bltest/tests/aes_gcm/ciphertext9
copy to cmd/bltest/tests/aes_gcm/ciphertext11
diff --git a/cmd/bltest/tests/aes_gcm/ciphertext10 b/cmd/bltest/tests/aes_gcm/ciphertext13
copy from cmd/bltest/tests/aes_gcm/ciphertext10
copy to cmd/bltest/tests/aes_gcm/ciphertext13
diff --git a/cmd/bltest/tests/aes_gcm/ciphertext11 b/cmd/bltest/tests/aes_gcm/ciphertext14
copy from cmd/bltest/tests/aes_gcm/ciphertext11
copy to cmd/bltest/tests/aes_gcm/ciphertext14
diff --git a/cmd/bltest/tests/aes_gcm/ciphertext12 b/cmd/bltest/tests/aes_gcm/ciphertext15
rename from cmd/bltest/tests/aes_gcm/ciphertext12
rename to cmd/bltest/tests/aes_gcm/ciphertext15
diff --git a/cmd/bltest/tests/aes_gcm/ciphertext13 b/cmd/bltest/tests/aes_gcm/ciphertext16
copy from cmd/bltest/tests/aes_gcm/ciphertext13
copy to cmd/bltest/tests/aes_gcm/ciphertext16
diff --git a/cmd/bltest/tests/aes_gcm/ciphertext14 b/cmd/bltest/tests/aes_gcm/ciphertext17
copy from cmd/bltest/tests/aes_gcm/ciphertext14
copy to cmd/bltest/tests/aes_gcm/ciphertext17
diff --git a/cmd/bltest/tests/aes_gcm/ciphertext1 b/cmd/bltest/tests/aes_gcm/ciphertext2
copy from cmd/bltest/tests/aes_gcm/ciphertext1
copy to cmd/bltest/tests/aes_gcm/ciphertext2
diff --git a/cmd/bltest/tests/aes_gcm/ciphertext2 b/cmd/bltest/tests/aes_gcm/ciphertext3
copy from cmd/bltest/tests/aes_gcm/ciphertext2
copy to cmd/bltest/tests/aes_gcm/ciphertext3
diff --git a/cmd/bltest/tests/aes_gcm/ciphertext3 b/cmd/bltest/tests/aes_gcm/ciphertext4
copy from cmd/bltest/tests/aes_gcm/ciphertext3
copy to cmd/bltest/tests/aes_gcm/ciphertext4
diff --git a/cmd/bltest/tests/aes_gcm/ciphertext4 b/cmd/bltest/tests/aes_gcm/ciphertext5
copy from cmd/bltest/tests/aes_gcm/ciphertext4
copy to cmd/bltest/tests/aes_gcm/ciphertext5
diff --git a/cmd/bltest/tests/aes_gcm/ciphertext5 b/cmd/bltest/tests/aes_gcm/ciphertext7
copy from cmd/bltest/tests/aes_gcm/ciphertext5
copy to cmd/bltest/tests/aes_gcm/ciphertext7
diff --git a/cmd/bltest/tests/aes_gcm/ciphertext6 b/cmd/bltest/tests/aes_gcm/ciphertext8
rename from cmd/bltest/tests/aes_gcm/ciphertext6
rename to cmd/bltest/tests/aes_gcm/ciphertext8
diff --git a/cmd/bltest/tests/aes_gcm/ciphertext7 b/cmd/bltest/tests/aes_gcm/ciphertext9
copy from cmd/bltest/tests/aes_gcm/ciphertext7
copy to cmd/bltest/tests/aes_gcm/ciphertext9
diff --git a/cmd/bltest/tests/aes_gcm/hex.c b/cmd/bltest/tests/aes_gcm/hex.c
new file mode 100644
--- /dev/null
+++ b/cmd/bltest/tests/aes_gcm/hex.c
@@ -0,0 +1,68 @@
+#include <unistd.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+int tohex(int c)
+{
+   if ((c >= '0') && (c <= '9')) {
+        return c - '0';
+   }
+   if ((c >= 'a') && (c <= 'f')) {
+        return c - 'a' + 10;
+   }
+   if ((c >= 'A') && (c <= 'F')) {
+        return c - 'A' + 10;
+   }
+   return 0;
+}
+
+int isspace(int c)
+{
+   if (c <= ' ') return 1;
+   if (c == '\n') return 1;
+   if (c == '\t') return 1;
+   if (c == ':') return 1;
+   if (c == ';') return 1;
+   if (c == ',') return 1;
+   return 0;
+}
+
+void verify_nibble(int nibble, int current)
+{
+   if (nibble != 0) {
+        fprintf(stderr,"count mismatch %d (nibbles=0x%x)\n",nibble,current);
+        fflush(stderr);
+   }
+}
+
+int main(int argc, char **argv)
+{
+   int c;
+   int current = 0;
+   int nibble = 0;
+   int skip = 0;
+
+   if (argv[1]) {
+        skip = atoi(argv[1]);
+   }
+
+#define NIBBLE_COUNT 2
+   while ((c=getchar()) != EOF) {
+      if (isspace(c)) {
+        verify_nibble(nibble,current);
+        continue;
+      }
+      if (skip) {
+        skip--;
+        continue;
+      }
+      current = current << 4 | tohex(c);
+      nibble++;
+      if (nibble == NIBBLE_COUNT) {
+        putchar(current);
+        nibble = 0;
+        current = 0;
+      }
+   }
+   return 0;
+}
diff --git a/cmd/bltest/tests/aes_gcm/iv0 b/cmd/bltest/tests/aes_gcm/iv1
rename from cmd/bltest/tests/aes_gcm/iv0
rename to cmd/bltest/tests/aes_gcm/iv1
diff --git a/cmd/bltest/tests/aes_gcm/iv8 b/cmd/bltest/tests/aes_gcm/iv10
copy from cmd/bltest/tests/aes_gcm/iv8
copy to cmd/bltest/tests/aes_gcm/iv10
diff --git a/cmd/bltest/tests/aes_gcm/iv9 b/cmd/bltest/tests/aes_gcm/iv11
copy from cmd/bltest/tests/aes_gcm/iv9
copy to cmd/bltest/tests/aes_gcm/iv11
diff --git a/cmd/bltest/tests/aes_gcm/iv10 b/cmd/bltest/tests/aes_gcm/iv13
copy from cmd/bltest/tests/aes_gcm/iv10
copy to cmd/bltest/tests/aes_gcm/iv13
diff --git a/cmd/bltest/tests/aes_gcm/iv11 b/cmd/bltest/tests/aes_gcm/iv14
copy from cmd/bltest/tests/aes_gcm/iv11
copy to cmd/bltest/tests/aes_gcm/iv14
diff --git a/cmd/bltest/tests/aes_gcm/iv12 b/cmd/bltest/tests/aes_gcm/iv15
rename from cmd/bltest/tests/aes_gcm/iv12
rename to cmd/bltest/tests/aes_gcm/iv15
diff --git a/cmd/bltest/tests/aes_gcm/iv13 b/cmd/bltest/tests/aes_gcm/iv16
copy from cmd/bltest/tests/aes_gcm/iv13
copy to cmd/bltest/tests/aes_gcm/iv16
diff --git a/cmd/bltest/tests/aes_gcm/iv14 b/cmd/bltest/tests/aes_gcm/iv17
copy from cmd/bltest/tests/aes_gcm/iv14
copy to cmd/bltest/tests/aes_gcm/iv17
diff --git a/cmd/bltest/tests/aes_gcm/iv1 b/cmd/bltest/tests/aes_gcm/iv2
copy from cmd/bltest/tests/aes_gcm/iv1
copy to cmd/bltest/tests/aes_gcm/iv2
diff --git a/cmd/bltest/tests/aes_gcm/iv2 b/cmd/bltest/tests/aes_gcm/iv3
copy from cmd/bltest/tests/aes_gcm/iv2
copy to cmd/bltest/tests/aes_gcm/iv3
diff --git a/cmd/bltest/tests/aes_gcm/iv3 b/cmd/bltest/tests/aes_gcm/iv4
copy from cmd/bltest/tests/aes_gcm/iv3
copy to cmd/bltest/tests/aes_gcm/iv4
diff --git a/cmd/bltest/tests/aes_gcm/iv4 b/cmd/bltest/tests/aes_gcm/iv5
copy from cmd/bltest/tests/aes_gcm/iv4
copy to cmd/bltest/tests/aes_gcm/iv5
diff --git a/cmd/bltest/tests/aes_gcm/iv5 b/cmd/bltest/tests/aes_gcm/iv7
copy from cmd/bltest/tests/aes_gcm/iv5
copy to cmd/bltest/tests/aes_gcm/iv7
diff --git a/cmd/bltest/tests/aes_gcm/iv6 b/cmd/bltest/tests/aes_gcm/iv8
rename from cmd/bltest/tests/aes_gcm/iv6
rename to cmd/bltest/tests/aes_gcm/iv8
diff --git a/cmd/bltest/tests/aes_gcm/iv7 b/cmd/bltest/tests/aes_gcm/iv9
copy from cmd/bltest/tests/aes_gcm/iv7
copy to cmd/bltest/tests/aes_gcm/iv9
diff --git a/cmd/bltest/tests/aes_gcm/key0 b/cmd/bltest/tests/aes_gcm/key1
rename from cmd/bltest/tests/aes_gcm/key0
rename to cmd/bltest/tests/aes_gcm/key1
diff --git a/cmd/bltest/tests/aes_gcm/key8 b/cmd/bltest/tests/aes_gcm/key10
copy from cmd/bltest/tests/aes_gcm/key8
copy to cmd/bltest/tests/aes_gcm/key10
diff --git a/cmd/bltest/tests/aes_gcm/key9 b/cmd/bltest/tests/aes_gcm/key11
copy from cmd/bltest/tests/aes_gcm/key9
copy to cmd/bltest/tests/aes_gcm/key11
diff --git a/cmd/bltest/tests/aes_gcm/key10 b/cmd/bltest/tests/aes_gcm/key13
copy from cmd/bltest/tests/aes_gcm/key10
copy to cmd/bltest/tests/aes_gcm/key13
diff --git a/cmd/bltest/tests/aes_gcm/key11 b/cmd/bltest/tests/aes_gcm/key14
copy from cmd/bltest/tests/aes_gcm/key11
copy to cmd/bltest/tests/aes_gcm/key14
diff --git a/cmd/bltest/tests/aes_gcm/key12 b/cmd/bltest/tests/aes_gcm/key15
rename from cmd/bltest/tests/aes_gcm/key12
rename to cmd/bltest/tests/aes_gcm/key15
diff --git a/cmd/bltest/tests/aes_gcm/key13 b/cmd/bltest/tests/aes_gcm/key16
copy from cmd/bltest/tests/aes_gcm/key13
copy to cmd/bltest/tests/aes_gcm/key16
diff --git a/cmd/bltest/tests/aes_gcm/key14 b/cmd/bltest/tests/aes_gcm/key17
copy from cmd/bltest/tests/aes_gcm/key14
copy to cmd/bltest/tests/aes_gcm/key17
diff --git a/cmd/bltest/tests/aes_gcm/key1 b/cmd/bltest/tests/aes_gcm/key2
copy from cmd/bltest/tests/aes_gcm/key1
copy to cmd/bltest/tests/aes_gcm/key2
diff --git a/cmd/bltest/tests/aes_gcm/key2 b/cmd/bltest/tests/aes_gcm/key3
copy from cmd/bltest/tests/aes_gcm/key2
copy to cmd/bltest/tests/aes_gcm/key3
diff --git a/cmd/bltest/tests/aes_gcm/key3 b/cmd/bltest/tests/aes_gcm/key4
copy from cmd/bltest/tests/aes_gcm/key3
copy to cmd/bltest/tests/aes_gcm/key4
diff --git a/cmd/bltest/tests/aes_gcm/key4 b/cmd/bltest/tests/aes_gcm/key5
copy from cmd/bltest/tests/aes_gcm/key4
copy to cmd/bltest/tests/aes_gcm/key5
diff --git a/cmd/bltest/tests/aes_gcm/key5 b/cmd/bltest/tests/aes_gcm/key7
copy from cmd/bltest/tests/aes_gcm/key5
copy to cmd/bltest/tests/aes_gcm/key7
diff --git a/cmd/bltest/tests/aes_gcm/key6 b/cmd/bltest/tests/aes_gcm/key8
rename from cmd/bltest/tests/aes_gcm/key6
rename to cmd/bltest/tests/aes_gcm/key8
diff --git a/cmd/bltest/tests/aes_gcm/key7 b/cmd/bltest/tests/aes_gcm/key9
copy from cmd/bltest/tests/aes_gcm/key7
copy to cmd/bltest/tests/aes_gcm/key9
diff --git a/cmd/bltest/tests/aes_gcm/mktst.sh b/cmd/bltest/tests/aes_gcm/mktst.sh
--- a/cmd/bltest/tests/aes_gcm/mktst.sh
+++ b/cmd/bltest/tests/aes_gcm/mktst.sh
@@ -1,10 +1,10 @@
 #!/bin/sh
-for i in 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14
+for i in 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17
 do
     file="test$i.txt"
     grep K= $file | sed -e 's;K=;;' | hex > key$i
     grep IV=  $file | sed -e 's;IV=;;' | hex > iv$i
     grep "C="  $file | sed -e 's;C=;;' | hex > ciphertext$i.bin
     grep "P="  $file | sed -e 's;P=;;' | hex  > plaintext$i
     grep "A="  $file | sed -e 's;A=;;' | hex  > aad$i
     grep "T="  $file | sed -e 's;T=;;' | hex  >> ciphertext$i.bin
diff --git a/cmd/bltest/tests/aes_gcm/numtests b/cmd/bltest/tests/aes_gcm/numtests
--- a/cmd/bltest/tests/aes_gcm/numtests
+++ b/cmd/bltest/tests/aes_gcm/numtests
@@ -1,1 +1,1 @@
-15
+18
diff --git a/cmd/bltest/tests/aes_gcm/plaintext0 b/cmd/bltest/tests/aes_gcm/plaintext1
rename from cmd/bltest/tests/aes_gcm/plaintext0
rename to cmd/bltest/tests/aes_gcm/plaintext1
diff --git a/cmd/bltest/tests/aes_gcm/plaintext8 b/cmd/bltest/tests/aes_gcm/plaintext10
copy from cmd/bltest/tests/aes_gcm/plaintext8
copy to cmd/bltest/tests/aes_gcm/plaintext10
diff --git a/cmd/bltest/tests/aes_gcm/plaintext9 b/cmd/bltest/tests/aes_gcm/plaintext11
copy from cmd/bltest/tests/aes_gcm/plaintext9
copy to cmd/bltest/tests/aes_gcm/plaintext11
diff --git a/cmd/bltest/tests/aes_gcm/plaintext10 b/cmd/bltest/tests/aes_gcm/plaintext13
copy from cmd/bltest/tests/aes_gcm/plaintext10
copy to cmd/bltest/tests/aes_gcm/plaintext13
diff --git a/cmd/bltest/tests/aes_gcm/plaintext11 b/cmd/bltest/tests/aes_gcm/plaintext14
copy from cmd/bltest/tests/aes_gcm/plaintext11
copy to cmd/bltest/tests/aes_gcm/plaintext14
diff --git a/cmd/bltest/tests/aes_gcm/plaintext12 b/cmd/bltest/tests/aes_gcm/plaintext15
rename from cmd/bltest/tests/aes_gcm/plaintext12
rename to cmd/bltest/tests/aes_gcm/plaintext15
diff --git a/cmd/bltest/tests/aes_gcm/plaintext13 b/cmd/bltest/tests/aes_gcm/plaintext16
copy from cmd/bltest/tests/aes_gcm/plaintext13
copy to cmd/bltest/tests/aes_gcm/plaintext16
diff --git a/cmd/bltest/tests/aes_gcm/plaintext14 b/cmd/bltest/tests/aes_gcm/plaintext17
copy from cmd/bltest/tests/aes_gcm/plaintext14
copy to cmd/bltest/tests/aes_gcm/plaintext17
diff --git a/cmd/bltest/tests/aes_gcm/plaintext1 b/cmd/bltest/tests/aes_gcm/plaintext2
copy from cmd/bltest/tests/aes_gcm/plaintext1
copy to cmd/bltest/tests/aes_gcm/plaintext2
diff --git a/cmd/bltest/tests/aes_gcm/plaintext2 b/cmd/bltest/tests/aes_gcm/plaintext3
copy from cmd/bltest/tests/aes_gcm/plaintext2
copy to cmd/bltest/tests/aes_gcm/plaintext3
diff --git a/cmd/bltest/tests/aes_gcm/plaintext3 b/cmd/bltest/tests/aes_gcm/plaintext4
copy from cmd/bltest/tests/aes_gcm/plaintext3
copy to cmd/bltest/tests/aes_gcm/plaintext4
diff --git a/cmd/bltest/tests/aes_gcm/plaintext4 b/cmd/bltest/tests/aes_gcm/plaintext5
copy from cmd/bltest/tests/aes_gcm/plaintext4
copy to cmd/bltest/tests/aes_gcm/plaintext5
diff --git a/cmd/bltest/tests/aes_gcm/plaintext5 b/cmd/bltest/tests/aes_gcm/plaintext7
copy from cmd/bltest/tests/aes_gcm/plaintext5
copy to cmd/bltest/tests/aes_gcm/plaintext7
diff --git a/cmd/bltest/tests/aes_gcm/plaintext6 b/cmd/bltest/tests/aes_gcm/plaintext8
rename from cmd/bltest/tests/aes_gcm/plaintext6
rename to cmd/bltest/tests/aes_gcm/plaintext8
diff --git a/cmd/bltest/tests/aes_gcm/plaintext7 b/cmd/bltest/tests/aes_gcm/plaintext9
copy from cmd/bltest/tests/aes_gcm/plaintext7
copy to cmd/bltest/tests/aes_gcm/plaintext9
diff --git a/cmd/bltest/tests/aes_gcm/test8.txt b/cmd/bltest/tests/aes_gcm/test10.txt
copy from cmd/bltest/tests/aes_gcm/test8.txt
copy to cmd/bltest/tests/aes_gcm/test10.txt
diff --git a/cmd/bltest/tests/aes_gcm/test9.txt b/cmd/bltest/tests/aes_gcm/test11.txt
copy from cmd/bltest/tests/aes_gcm/test9.txt
copy to cmd/bltest/tests/aes_gcm/test11.txt
diff --git a/cmd/bltest/tests/aes_gcm/test10.txt b/cmd/bltest/tests/aes_gcm/test13.txt
copy from cmd/bltest/tests/aes_gcm/test10.txt
copy to cmd/bltest/tests/aes_gcm/test13.txt
diff --git a/cmd/bltest/tests/aes_gcm/test11.txt b/cmd/bltest/tests/aes_gcm/test14.txt
copy from cmd/bltest/tests/aes_gcm/test11.txt
copy to cmd/bltest/tests/aes_gcm/test14.txt
diff --git a/cmd/bltest/tests/aes_gcm/test13.txt b/cmd/bltest/tests/aes_gcm/test16.txt
copy from cmd/bltest/tests/aes_gcm/test13.txt
copy to cmd/bltest/tests/aes_gcm/test16.txt
diff --git a/cmd/bltest/tests/aes_gcm/test14.txt b/cmd/bltest/tests/aes_gcm/test17.txt
copy from cmd/bltest/tests/aes_gcm/test14.txt
copy to cmd/bltest/tests/aes_gcm/test17.txt
diff --git a/cmd/bltest/tests/aes_gcm/test1.txt b/cmd/bltest/tests/aes_gcm/test2.txt
copy from cmd/bltest/tests/aes_gcm/test1.txt
copy to cmd/bltest/tests/aes_gcm/test2.txt
diff --git a/cmd/bltest/tests/aes_gcm/test2.txt b/cmd/bltest/tests/aes_gcm/test3.txt
copy from cmd/bltest/tests/aes_gcm/test2.txt
copy to cmd/bltest/tests/aes_gcm/test3.txt
diff --git a/cmd/bltest/tests/aes_gcm/test3.txt b/cmd/bltest/tests/aes_gcm/test4.txt
copy from cmd/bltest/tests/aes_gcm/test3.txt
copy to cmd/bltest/tests/aes_gcm/test4.txt
diff --git a/cmd/bltest/tests/aes_gcm/test4.txt b/cmd/bltest/tests/aes_gcm/test5.txt
copy from cmd/bltest/tests/aes_gcm/test4.txt
copy to cmd/bltest/tests/aes_gcm/test5.txt
diff --git a/cmd/bltest/tests/aes_gcm/test5.txt b/cmd/bltest/tests/aes_gcm/test7.txt
copy from cmd/bltest/tests/aes_gcm/test5.txt
copy to cmd/bltest/tests/aes_gcm/test7.txt
diff --git a/cmd/bltest/tests/aes_gcm/test7.txt b/cmd/bltest/tests/aes_gcm/test9.txt
copy from cmd/bltest/tests/aes_gcm/test7.txt
copy to cmd/bltest/tests/aes_gcm/test9.txt
diff --git a/cmd/bltest/tests/aes_gcm/test_source.txt b/cmd/bltest/tests/aes_gcm/test_source.txt
--- a/cmd/bltest/tests/aes_gcm/test_source.txt
+++ b/cmd/bltest/tests/aes_gcm/test_source.txt
@@ -148,25 +148,26 @@ X4=480c00db2679065a7ed2f771a53acacd
 X5=1c1ae3c355e2214466a9923d2ba6ab35
 X6=0694c6f16bb0275a48891d06590344b0
 len(A)||len(C)=00000000000000a000000000000001e0
 GHASH(H,A,C)=1c5afe9760d3932f3c9a878aac3dc3de
 C=8ce24998625615b603a033aca13fb894be9112a5c3a211a8ba262a3cca7e2ca701e4a9a4fba43c90ccdcb281d48c7c6fd62875d2aca417034c34aee5
 T=619cc5aefffe0bfa462af43c1699d050
 
 test="Test Case 7"
-K=000000000000000000000000000000000000000000000000000000000000000000000000
+K=000000000000000000000000000000000000000000000000
 P=
-IV=aae06992acbf52a3e8f4a96ec9300bd7
-H=00000000000000000000000000000001
-Y0=cd33b28ac773f74ba00ed1f312572435
-E(K,Y0)=00000000000000000000000000000000
+IV=000000000000000000000000
+H=aae06992acbf52a3e8f4a96ec9300bd7
+Y0=00000000000000000000000000000001
+E(K,Y0)=cd33b28ac773f74ba00ed1f312572435
+len(A)||len(C)=00000000000000000000000000000000
 GHASH(H,A,C)=00000000000000000000000000000000
 C=
-T=cd33b28ac773f74ba00ed1f31257243
+T=cd33b28ac773f74ba00ed1f312572435
 
 test="Test Case 8"
 K=000000000000000000000000000000000000000000000000
 P=00000000000000000000000000000000
 IV=000000000000000000000000
 H=aae06992acbf52a3e8f4a96ec9300bd7
 Y0=00000000000000000000000000000001
 E(K,Y0)=cd33b28ac773f74ba00ed1f312572435
diff --git a/tests/cipher/cipher.txt b/tests/cipher/cipher.txt
--- a/tests/cipher/cipher.txt
+++ b/tests/cipher/cipher.txt
@@ -15,16 +15,19 @@
 	0	des3_ecb_-E	DES3_ECB_Encrypt
 	0	des3_ecb_-D	DES3_ECB_Decrypt
 	0	des3_cbc_-E	DES3_CBC_Encrypt
 	0	des3_cbc_-D	DES3_CBC_Decrypt
 	0	aes_ecb_-E	AES_ECB_Encrypt
 	0	aes_ecb_-D	AES_ECB_Decrypt
 	0	aes_cbc_-E	AES_CBC_Encrypt
 	0	aes_cbc_-D	AES_CBC_Decrypt
+	0	aes_ctr		AES_CTR
+	0	aes_cts		AES_CTS
+	0	aes_gcm		AES_GCM
 	0	camellia_ecb_-E	Camellia_ECB_Encrypt
 	0	camellia_ecb_-D	Camellia_ECB_Decrypt
 	0	camellia_cbc_-E	Camellia_CBC_Encrypt
 	0	camellia_cbc_-D	Camellia_CBC_Decrypt
 	0	seed_ecb_-E	SEED_ECB_Encrypt
 	0	seed_ecb_-D	SEED_ECB_Decrypt
 	0	seed_cbc_-E	SEED_CBC_Encrypt
 	0	seed_cbc_-D	SEED_CBC_Decrypt
diff --git a/tests/cipher/dsa.txt b/tests/cipher/dsa.txt
--- a/tests/cipher/dsa.txt
+++ b/tests/cipher/dsa.txt
@@ -4,8 +4,10 @@
 #
 # This file enables test coverage of the dsa performance tests
 #
 #
 # mode     keysize   bufsize   repetitions cxrepetitions
   dsa        64         20         200         5
   dsa        96         20         200         3
   dsa       128         20         200         3
+  dsa       256         20         200         3
+  dsa       384         20         200         3
diff --git a/tests/cipher/symmkey.txt b/tests/cipher/symmkey.txt
--- a/tests/cipher/symmkey.txt
+++ b/tests/cipher/symmkey.txt
@@ -19,14 +19,18 @@
  rc4          5       8192       10000        100000
  rc4          8       8192       10000        100000
  rc4         16       8192       10000        100000
  rc4         24       8192       10000        100000
  aes_ecb     16       8192       10000        100000
  aes_cbc     16       8192       10000        100000
  aes_ecb     32       8192       10000        100000
  aes_cbc     32       8192       10000        100000
+ aes_ctr     16       8192       10000        100000
+ aes_ctr     32       8192       10000        100000
+ aes_gcm     16       8192       10000        100000
+ aes_gcm     32       8192       10000        100000
  camellia_ecb     16       8192       10000        100000
  camellia_cbc     16       8192       10000        100000
  camellia_ecb     32       8192       10000        100000
  camellia_cbc     32       8192       10000        100000
  seed_ecb    16       8192       10000        100000
  seed_cbc    16       8192       10000        100000
