diff -up ./mozilla/security/nss/lib/softoken/sdb.c.1044666 ./mozilla/security/nss/lib/softoken/sdb.c
--- ./mozilla/security/nss/lib/softoken/sdb.c.1044666	2014-03-21 11:33:50.567190516 -0700
+++ ./mozilla/security/nss/lib/softoken/sdb.c	2014-03-21 11:33:50.675191304 -0700
@@ -2125,7 +2125,17 @@ s_open(const char *directory, const char
 
     /* how long does it take to test for a non-existant file in our working
      * directory? Allows us to test if we may be on a network file system */
-    accessOps = sdb_measureAccess(directory);
+    accessOps = 1;
+    {
+        char *env;
+        env = PR_GetEnv("NSS_SDB_USE_CACHE");
+        /* If the environment variable is set to yes or no, sdb_init() will
+         * ignore the value of accessOps, and we can skip the measuring. */
+        if (!env || ((PORT_Strcasecmp(env, "no") != 0) &&
+                     (PORT_Strcasecmp(env, "yes") != 0))){
+           accessOps = sdb_measureAccess(directory);
+        }
+    }
 
     /*
      * open the cert data base
